### PR TITLE
Redo wait() API.

### PR DESCRIPTION
**Pull request by ipuustin**
The idea is that the shim implementations would need to do less thread management.

Fixes: #67 

@cpuguy83 would this approach make sense? It should maybe be part of a wider API change if it is needed.

### CODE REVIEW

**Review by Mossaka — APPROVED** [review]
lgtm!

**Comment by Mossaka** [comment]
@cpuguy83 can you take a look?

**Comment by cpuguy83** [comment]
Thanks for taking a look!

I'm not sure I understand how this changes the interfaces/call structure.
Instead of a tx we wrap a tx but still need to deal with it effectively the same way.
Maybe I overlooked something?

**Comment by ipuustin** [comment]
> I'm not sure I understand how this changes the interfaces/call structure. Instead of a tx we wrap a tx but still need to deal with it effectively the same way. Maybe I overlooked something?

Well, the idea is just that the shims don't need to understand how the thread management should be done, because that's handled by the wrapper -- I understood that was what the original issue was after. Can you elaborate a bit what kind of API you would like to see and I can take another shot at this. :-)

**Comment by cpuguy83** [comment]
@ipuustin Ok yes I totally just brain farted on this yesterday.
Design looks great.

File: crates/containerd-shim-wasm/src/sandbox/instance.rs [hunk 1]
```diff
@@ -112,10 +112,34 @@ pub trait Instance {
...
     /// wait for the instance to exit
-    /// The sender is used to send the exit code and time back to the caller
-    /// Ideally this would just be a blocking call with a normal result, however
-    /// because of how this is called from a thread it causes issues with lifetimes of the trait implementer.
-    fn wait(&self, send: Sender<(u32, DateTime<Utc>)>) -> Result<(), Error>;
+    /// The waiter is used to send the exit code and time back to the caller
+    fn wait(&self, waiter: &Wait) -> Result<(), Error>;
+}
+
+pub struct Wait {
```
[the remark this answers is not in the document]
Comment (ipuustin) [reply crates/containerd-shim-wasm/src/sandbox/instance.rs:125]: I added docs and also tried to think of a better name for the function.

**Comment by danbugs** [comment]
~might want to update the `README.md`'s mention of `wait`: https://github.com/containerd/runwasi/blob/fff59017205828e1835ea7f699b7a0c5e7b924be/README.md?plain=1#L41

**Comment by cpuguy83** [comment]
Looks like this needs a rebase.

**Review by cpuguy83 — APPROVED** [review]

**Pull request merged**